packages/libs/error-stack/src/hook/context.rs

```diff
@@ -0,0 +1,366 @@
+    ///
+    /// In most situations this functions isn't needed, as it transparently casts between different
+    /// partitions of the storage. Only hooks that share storage with hooks of different types
+    /// should need to use this function.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// # // we only test with Rust 1.65, which means that `render()` is unused on earlier version
+    /// # #![cfg_attr(not(rust_1_65), allow(dead_code, unused_variables, unused_imports))]
+    /// use std::io::ErrorKind;
+    ///
+    /// use error_stack::Report;
+    ///
+    /// struct Warning(&'static str);
+    /// struct Error(&'static str);
+    ///
+    /// Report::install_debug_hook::<Error>(|Error(frame), context| {
+    ///     let idx = context.increment_counter() + 1;
+    ///
+    ///     context.push_body(format!("[{idx}] [ERROR] {frame}"));
+    /// });
+    /// Report::install_debug_hook::<Warning>(|Warning(frame), context| {
+    ///     // We want to share the same counter with `Error`, so that we're able to have
+    ///     // a global counter to keep track of all errors and warnings in order, this means
+    ///     // we need to access the storage of `Error` using `cast()`.
+    ///     let context = context.cast::<Error>();
+    ///     let idx = context.increment_counter() + 1;
+    ///     context.push_body(format!("[{idx}] [WARN] {frame}"))
+    /// });
+    ///
+    /// let report = Report::new(std::io::Error::from(ErrorKind::InvalidInput))
+    ///     .attach(Error("unable to reach remote host"))
+    ///     .attach(Warning("disk nearly full"))
+    ///     .attach(Error("cannot resolve example.com: unknown host"));
+    ///
+    /// # owo_colors::set_override(true);
+    /// # fn render(value: String) -> String {
+    /// #     let backtrace = regex::Regex::new(r"backtrace no\. (\d+)\n(?:  .*\n)*  .*").unwrap();
+    /// #     let backtrace_info = regex::Regex::new(r"backtrace( with (\d+) frames)? \((\d+)\)").unwrap();
+    /// #
+    /// #     let value = backtrace.replace_all(&value, "backtrace no. $1\n  [redacted]");
+    /// #     let value = backtrace_info.replace_all(value.as_ref(), "backtrace ($3)");
+    /// #
+    /// #     ansi_to_html::convert_escaped(value.as_ref()).unwrap()
+    /// # }
+    /// #
+    /// # #[cfg(rust_1_65)]
+    /// # expect_test::expect_file![concat!(env!("CARGO_MANIFEST_DIR"), "/tests/snapshots/doc/fmt__hookcontext_cast.snap")].assert_eq(&render(format!("{report:?}")));
+    /// #
+    /// println!("{report:?}");
+    /// ```
+    ///
+    /// <pre>
+    #[doc = include_str!(concat!(env!("CARGO_MANIFEST_DIR"), "/tests/snapshots/doc/fmt__hookcontext_cast.snap"))]
+    /// </pre>
+    #[must_use]
+    pub fn cast<U>(&mut self) -> &mut HookContext<Extra, U> {
+        // SAFETY: `HookContext` is marked as repr(transparent) and the changed generic is only used
+        // inside of the `PhantomData`
+        unsafe { &mut *(self as *mut Self).cast::<HookContext<Extra, U>>() }
+    }
```

Comment:
I take the opportunity to ask this question as we touched this code: Do you see a way to remove the `unsafe` here? I know, why we have, I'm just wondering if there is a way to avoid it (this question is unrelated to this PR).

Reply:
I don't think so, if we would instead transmute `mut self` instead we could use RFC 2528 may be of help.